src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.Snapshotter.cs

```diff
@@ -0,0 +1,250 @@
+            private readonly bool _useMethodSignatureParameterTypes;
+            private readonly Symbol _symbol;
+            private readonly ImmutableDictionary<BoundExpression, TypeWithState> _methodGroupReceiverMapOpt;
+            private readonly VisitResult _visitResult;
+            private readonly VisitResult _currentConditionalReceiverVisitResult;
+            private readonly ImmutableDictionary<object, PlaceholderLocal> _placeholderLocalsOpt;
+            private readonly int _lastConditionalAccessSlot;
+
+            private Snapshot(
+                VariableState variableState,
+                bool useMethodSignatureParameterTypes,
+                Symbol symbol,
+                ImmutableDictionary<BoundExpression, TypeWithState> methodGroupReceiverMapOpt,
+                VisitResult visitResult,
+                VisitResult currentConditionalReceiverVisitResult,
+                ImmutableDictionary<object, PlaceholderLocal> placeholderLocalsOpt,
+                int lastConditionalAccessSlot)
+            {
+                _variableState = variableState;
+                _useMethodSignatureParameterTypes = useMethodSignatureParameterTypes;
+                _symbol = symbol;
+                _methodGroupReceiverMapOpt = methodGroupReceiverMapOpt;
+                _visitResult = visitResult;
+                _currentConditionalReceiverVisitResult = currentConditionalReceiverVisitResult;
+                _placeholderLocalsOpt = placeholderLocalsOpt;
+                _lastConditionalAccessSlot = lastConditionalAccessSlot;
+            }
+
+            internal static Snapshot SnapshotWalker(NullableWalker walker) =>
+                new Snapshot(
+                    walker._snapshotter.GetVariableState(walker.State),
+                    walker._useMethodSignatureParameterTypes,
+                    walker._symbol,
+                    walker._methodGroupReceiverMapOpt,
+                    walker._visitResult,
+                    walker._currentConditionalReceiverVisitResult,
+                    walker._placeholderLocalsOpt,
+                    walker._lastConditionalAccessSlot);
+
+            internal NullableWalker RestoreFromCheckpoint(CSharpCompilation compilation,
+                                                     BoundNode nodeToAnalyze,
+                                                     Binder binder,
+                                                     Dictionary<BoundExpression, (NullabilityInfo, TypeSymbol)> analyzedNullabilityMap)
+            {
+                return new NullableWalker(
+                    compilation,
+                    this._symbol,
+                    this._useMethodSignatureParameterTypes,
+                    this._symbol as MethodSymbol,
+                    nodeToAnalyze,
+                    binder,
+                    binder.Conversions,
+                    this._variableState,
+                    returnTypesOpt: default,
+                    analyzedNullabilityMap,
```

Comment:
>analyzedNullabilityMap [](start = 20, length = 22)

Is the walker going to add to the map or to look for something in it?

Reply:
It will add to the map. This is how nullability info on the speculative code will be conveyed to the speculative model.

---
In reply to: [288720277](https://github.com/dotnet/roslyn/pull/35850#discussion_r288720277) [](ancestors = 288720277)